### ISSUE TITLE
Serializing coroutine's state

### ISSUE THREAD

**Issue by abreslav** [Kotlin/kotlin-coroutines#28]
### Use cases

Server applications often implement protocols like
- register a user - send confirmation email - wait for confirmation - proceed with login
- receive a request from a user - place an order - do some work - send a notification on completion

These are often implemented as explicit state machines, but they would be very naturally written as coroutines:

```
protocol {
    val userData = showRegistrationForm() // suspension point
    db.enterData(userData)
    awaitEmailConfirmation(userData.userId) // suspension point
    db.allowLogin(userData.userId)
}
```
### Related reading
- http://swarmframework.org/ - distributed computations
- http://www.seaside.st/about/examples/task?_k=x47Kg2-M - multi-page workflows in Seaside
- https://ep2013.europython.eu/media/conference/slides/advanced-pickling-with-stackless-python-and-spickle.pdf
### Challenges

To allow such approach to scale, we'd need to serialize the state of coroutines to a database, so that it can be resumed on another node. There are the following issues with it:
- altering the code of a coroutine may change the state table of the state machine subtly, and the resumed code will break
- not all the data captured by a coroutine may be serializable
### Notable comments
- [Use `@Serializable` on the coroutine parameter to direct compiler checks](https://github.com/Kotlin/kotlin-coroutines/issues/28#issuecomment-222627470)

**Comment by abreslav** [comment]
> altering the code of a coroutine may change the state table of the state machine subtly, and the resumed code will break

A step toward a solution would be to make a hash of the coroutine's byte code available through some API. This way an application can at least detect that something's not compatible anymore.

This won't help it continue working, though :(

**Comment by abreslav** [comment]
> not all the data captured by a coroutine may be serializable

We could impose some discipline regarding what can be captured in such coroutines, i.e. in the closure, parameters and locals shared between states. This could be done either by the compiler or even by the runtime library if we provide enough means of introspection.

**Comment by mikehearn** [comment]
My current project serialises Quasar fibers for this exact reason. I will report back along the way as we gain experience with this style of programming. Right now the main issue is that Quasar pushes at the limits of our complexity budget (i.e. it can be glitchy and take time to debug what's going on).

**Comment by bsideup** [comment]
👍

**Comment by bsideup** [comment]
@abreslav we could also add Serializable as annotation on coroutine parameter to force compiler to make sure that everything inside a coroutine is serializable

**Comment by abreslav** [comment]
@bsideup That's definitely an option worth considering. We should think of what checks should be applied exactly

**Comment by MichaelRocks** [comment]
Another extremely useful case for coroutine state serialization is retrieving a result from another activity on Android or checking for a permission. In the former case calling code invokes `startActivityForResult()` and receives a result in `onActivityResult()`. It seems to be easy to implement this case using coroutines. However, the calling process may be terminated and restarted before receiving the result so we need to serialize the coroutine's state and restore it when the process is relaunched.

**Comment by abreslav** [comment]
@MichaelRocks How is such a use case (process is terminated before the result is received) handled without coroutines? Do we need to serialize something manually?

**Comment by MichaelRocks** [comment]
@abreslav When performing one of these system calls you have to specify an integer request code. Later Android passes this code to a callback so you can distinct this invocation from other ones. If you need some other state, and usually you do, you can persist the state in a system callback called `onSaveInstanceState()` and restore it after the process is relaunched.

 If coroutine state is serializable one can persist it across process restart and continue its execution in the new process.

**Comment by sanity** [comment]
Just to chime in from a user's perspective (I'm the author of the Swarm proposal, referenced above), while it would be nice if this mechanism is robust against code changes between when state is serialized and deserialized - I see that as a responsibility that would be handled at the library level, and so I wouldn't personally consider it a blocker for this feature.

Similarly, while it would be nice if the compiler could verify serializability, wouldn't that, for example, prevent the serialization of open classes or interfaces, since the serializability of subclasses cannot be assured?  Again, I would see this as a "nice to have" but not a "need to have", at least for the application I have in mind.

**Comment by abreslav** [comment]
@sanity We'll be happy to help you get started: if there's something small (and even dirty) that you need to kick the swarm work off, go ahead and tell us. We'll try to enable your use cases ASAP

**Comment by sanity** [comment]
Thank you @abreslav, we're still in the planning stages but will let you know ASAP.

**Comment by oravecz** [comment]
@mikehearn Did quasar end up getting serialization support?

**Comment by dzharkov** [comment]
Note that in current implementation all coroutines implicitly implement `java.io.Serializable` just as well as all common lambda classes do

**Comment by defHLT** [comment]
@dzharkov doesn't seems to be the case anymore:

` java.lang.ClassCastException: kotlinx.coroutines.experimental.StandaloneCoroutine cannot be cast to java.io.Serializable`

**Comment by dzharkov** [comment]
// cc @elizarov

**Comment by Restioson** [comment]
+1

Another use case is perhaps for game AI: you can write a coroutine which, say, waits for the player to come near, and toggles a door, and waits for them to attack it, upon which it becomes aggressive. Something like:

```
waitEvent(PlayerProximityEvent(10))
door.toggle() 
waitEvent(DamageEvent(player)) 
aggression = true
```
If the coroutine is restarted from where it left off, say, waiting for the player to attack, then all is well. However, if the coroutine is restarted upon deserialisation from the *beginning*, it will immediately toggle the door (as the player is near as the proximity event has already been triggered), leaving the door in the wrong state. state

**Comment by elizarov** [comment]
https://github.com/Kotlin/kotlinx.coroutines/issues/76

**Comment by elizarov** [comment]
@Restioson You can try to write a prototype using some 3rd party serialization framework like Kryo. The general idea, is that you have to implement your own coroutine context and provide your own suspending functions like `waitEvent`. The idea is that `waitEvent` implementation invokes `suspendCoroutine` and then stores the resulting `Continuation` reference in its own data-structure that lists all continuations and what evens they are waiting for. You can then serialize this data structure and it will get all waiting coroutines serialized, too, together with their state machines. Upon deserialization of this data structure you can continue running your game, scheduling coroutines (invoking `Continuation.resume`) as game events happen.

**Comment by Restioson** [comment]
Thanks so much! I also thought of having waitEvent wait on a future which would be shared between coroutines and fulfilled when an event is registered and then writing a custom builder to keep track of the Continuations.

**Comment by elizarov** [comment]
You don't need a custom builder. Builder does not have access to the state-machine of coroutines. It is only a suspending function like `suspendCoroutine` (that you could use in your `waitEvent` implementation) that exposes a reference to the coroutine state, which can be then serialized.

**Comment by elizarov** [comment]
The corresponding YouTrack issue for serializability of classes from Kotlin stdlib is: https://youtrack.jetbrains.com/issue/KT-18559 

No long-term stability of serialization is planned in the short term.

**Comment by greenbergjosh** [comment]
I just started looking at Kotlin specifically to see if it could do this sort of thing. I have been trying to figure this out in the .Net world for quite some time. Thought I had it when I found: https://blogs.msdn.microsoft.com/lucian/2016/04/20/async-workflow-2/

I eventually got something working based on that code, but then ran into an important issue that I felt would be beneficial to bring up early in this forum.  I use Roslyn to enable C# as a scripting language.  The yield from async stopped working when I called into a dynamic, async method (as opposed to a static async method in my code file).

An optimal implementation would combine JSR223 (https://discuss.kotlinlang.org/t/programmatic-way-to-run-kotlin-as-a-scripting-language-via-jsr-223-or-a-homegrown-api/1686) with the ability to serialize coroutines, and allow these scripted coroutines to be serialized in the same fashion as static coroutines.  

This would be the holy grail for cluster computing, long-running workflows, rule engines, etc.
